<style>
.toc {
  font-family: Avenir;
  font-size: 16px;
}
</style>
<div style="text-align:center;">
  <h1>Delivering Delight: Insights and Analytics for Food Delivery Operations</h1>
  <h2>A Comprehensive Dashboard for Tracking Orders, Drivers, and Customer Satisfaction in Real-Time</h2>
  <h5>By <a href="https://www.linkedin.com/in/chiraglakhanpal/">Chirag Lakhanpal</a></h5>
</div> 
<br>
<br>

<style>.toc {font-family: Avenir;font-size: 16px;}</style>
<h2>Table of Contents<span class="tocSkip"></span></h2>
<div class="toc">
    <ul class="toc-item">
        <li>
            <span>
                <a href="## Exordium" data-toc-modified-id="Exordium-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Exordium</a>
            </span>
        </li>
        <li>
            <span>
                <a href="#Notebook-Configuration" data-toc-modified-id="Notebook-Configuration-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Notebook Configuration</a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Packages" data-toc-modified-id="Warning-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Packages</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Warning" data-toc-modified-id="Warning-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Warning</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Random-seed" data-toc-modified-id="Random-seed-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Random seed</a>
                    </span>
                </li>
            </ul>
        </li>
        <li>
            <span>
                <a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Preprocessing</a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Loading-the-data" data-toc-modified-id="Loading-the-data-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Data Description</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Loading-the-data" data-toc-modified-id="Loading-the-data-3.1"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Loading the data</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#lookup" data-toc-modified-id="Loading-the-data-3.1"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Looking up Market ID to City Name</a>
                    </span>
                </li>                
                <li>
                    <span>
                        <a href="#Splitting-the-data" data-toc-modified-id="Splitting-the-data-3.2"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Splitting the data</a>
                    </span>
                </li>
               <li>
                    <span>
                        <a href="#Handling-identifiers" data-toc-modified-id="Handling-identifiers-3.4"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Handling identifiers</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Combining-the-training,-validation-and-test-data" data-toc-modified-id="Combining-the-training,-validation-and-test-data-3.4.1">
                                    <span class="toc-item-num">3.5.1&nbsp;&nbsp;</span>Combining the training, validation and test data
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-identifiers" data-toc-modified-id="Identifying-identifiers-3.4.2"><span class="toc-item-num">3.5.2&nbsp;&nbsp;</span>Identifying identifiers</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.5.3&nbsp;&nbsp;</span>Removing identifiers</a>
                            </span>
                        </li>
                    </ul>
                </li>
                <li>
                    <span>
                        <a href="#Handling-missing-data" data-toc-modified-id="Handling-missing-data-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Handling missing data</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Combining-the-training,-validation-and-test-data" data-toc-modified-id="Combining-the-training,-validation-and-test-data-3.6.1">
                                    <span class="toc-item-num">3.6.1&nbsp;&nbsp;</span>Combining the training, validation and test data
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-missing-values" data-toc-modified-id="Identifying-missing-values-3.6.2"><span class="toc-item-num">3.6.2&nbsp;&nbsp;</span>Identifying missing values</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Separating-the-training,-validation-and-test-data" data-toc-modified-id="Separating-the-training,-validation-and-test-data-3.6.3">
                                    <span class="toc-item-num">3.6.3&nbsp;&nbsp;</span>Separating the training, validation and test data
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Imputing-missing-values" data-toc-modified-id="Imputing-missing-values-3.6.4"><span class="toc-item-num">3.6.4&nbsp;&nbsp;</span>Imputing missing values</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Imputing-missing-values" data-toc-modified-id="Imputing-missing-values-3.6.4"><span class="toc-item-num">3.6.5&nbsp;&nbsp;</span>Droping missing values</a>
                            </span>
                        </li>                 
                    </ul>
                </li>
               <li>
                    <span>
                        <a href="#Handling-identifiers" data-toc-modified-id="Handling-identifiers-3.4"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Mapping data types</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Combining-the-training,-validation-and-test-data" data-toc-modified-id="Combining-the-training,-validation-and-test-data-3.4.1">
                                    <span class="toc-item-num">3.7.1&nbsp;&nbsp;</span>Combining the training, validation and test data
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-identifiers" data-toc-modified-id="Identifying-identifiers-3.4.2"><span class="toc-item-num">3.7.2&nbsp;&nbsp;</span>Numerical Features</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.7.3&nbsp;&nbsp;</span>TimeStamp Features</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.7.4&nbsp;&nbsp;</span>Categorical Features</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.7.5&nbsp;&nbsp;</span>Improving Readability</a>
                            </span>
                        </li>  
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.7.6&nbsp;&nbsp;</span>Separating the training, validation and test data</a>
                            </span>
                        </li>                        
                    </ul>
                </li>           
                <li>
                    <span>
                        <a href="#Handling-uncommon-features" data-toc-modified-id="Handling-uncommon-features-3.3"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Removing unnecessary features</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.8.1&nbsp;&nbsp;</span>Combining the training, validation and test data</a>
                            </span>
                        </li>
                    </ul>
                </li>
                <li>
                    <span>
                        <a href="#Handling-uncommon-features" data-toc-modified-id="Handling-uncommon-features-3.3"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>Creating a Duplicate Dataset for EDA and Modeling</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.9.1&nbsp;&nbsp;</span>Combining the training, validation and test data</a>
                            </span>
                        </li>
                    </ul>
                </li>
                <li>
                    <span>
                        <a href="#Encoding-the-data" data-toc-modified-id="Encoding-the-data-3.7"><span class="toc-item-num">3.10&nbsp;&nbsp;</span>Encoding the data</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Combining-the-training,-validation-and-test-data" data-toc-modified-id="Combining-the-training,-validation-and-test-data-3.7.1">
                                    <span class="toc-item-num">3.10.1&nbsp;&nbsp;</span>Combining the training, validation and test data
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-categorical-variables" data-toc-modified-id="Identifying-categorical-variables-3.7.2"><span class="toc-item-num">3.10.2&nbsp;&nbsp;</span>Identifying categorical variables</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Encoding-categorical-features" data-toc-modified-id="Encoding-categorical-features-3.7.3"><span class="toc-item-num">3.10.3&nbsp;&nbsp;</span>Encoding categorical features</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Separating-the-training,-validation-and-test-data" data-toc-modified-id="Separating-the-training,-validation-and-test-data-3.7.4">
                                    <span class="toc-item-num">3.10.4&nbsp;&nbsp;</span>Separating the training, validation and test data
                                </a>
                            </span>
                        </li>
                    </ul>
                </li>
                <li>
                    <span>
                        <a href="#Handling-uncommon-features" data-toc-modified-id="Handling-uncommon-features-3.3"><span class="toc-item-num">3.11&nbsp;&nbsp;</span>Handling date and time variable</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.11.1&nbsp;&nbsp;</span>Transforming date time variables</a>
                            </span>
                        </li>
                    </ul>
                </li>
               <li>
                    <span>
                        <a href="#Handling-identifiers" data-toc-modified-id="Handling-identifiers-3.4"><span class="toc-item-num">3.12&nbsp;&nbsp;</span>Feature Engineering</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Combining-the-training,-validation-and-test-data" data-toc-modified-id="Combining-the-training,-validation-and-test-data-3.4.1">
                                    <span class="toc-item-num">3.12.1&nbsp;&nbsp;</span>Combining the training, validation and test data
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-identifiers" data-toc-modified-id="Identifying-identifiers-3.4.2"><span class="toc-item-num">3.12.2&nbsp;&nbsp;</span>Target Variable (ETA)</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.3&nbsp;&nbsp;</span>Busy Rider Ratio (%)</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.4&nbsp;&nbsp;</span>Non-Preparation Duration</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.5&nbsp;&nbsp;</span>Day of the week</a>
                            </span>
                        </li>  
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.6&nbsp;&nbsp;</span>Average Order Price</a>
                            </span>
                        </li>                        
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.7&nbsp;&nbsp;</span>Time of the day </a>
                            </span>
                        </li>                        
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.8&nbsp;&nbsp;</span>Delivery Duration</a>
                            </span>
                        </li>                        
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.9&nbsp;&nbsp;</span>Trip Nature</a>
                            </span>
                        </li>                        
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.10&nbsp;&nbsp;</span>Delivery distance</a>
                            </span>
                        </li>                        
                        <li>
                            <span>
                                <a href="#Removing-identifiers" data-toc-modified-id="Removing-identifiers-3.4.3"><span class="toc-item-num">3.12.11&nbsp;&nbsp;</span>Order Density</a>
                            </span>
                        </li>                                              
                    </ul>
                </li>                         
                <li>
                    <span>
                        <a href="#Handling-uncommon-features" data-toc-modified-id="Handling-uncommon-features-3.3"><span class="toc-item-num">3.13&nbsp;&nbsp;</span>Outlier Detection and Treatment</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.13.1&nbsp;&nbsp;</span>Outliers Detection</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.13.2&nbsp;&nbsp;</span>Outlier Treatment</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.13.3&nbsp;&nbsp;</span>Outlier Rules for Dataset Ranges</a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Identifying-uncommon-features" data-toc-modified-id="Identifying-uncommon-features-3.3.1"><span class="toc-item-num">3.13.4&nbsp;&nbsp;</span>Defining Range</a>
                            </span>
                        </li>                  
                    </ul>
                </li>              
                <li>
                    <span>
                        <a href="#Scaling-the-data" data-toc-modified-id="Scaling-the-data-3.9"><span class="toc-item-num">3.14&nbsp;&nbsp;</span>Scaling the data</a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Standardization" data-toc-modified-id="Standardization-3.9.1"><span class="toc-item-num">3.14.1&nbsp;&nbsp;</span>Standardization</a>
                            </span>
                            <ul class="toc-item">
                            </ul>
                        </li>
                    </ul>
                </li>
            </ul>
        </li>
        <li>
            <span>
                <a href="#Hyperparameter-Tuning" data-toc-modified-id="Hyperparameter-Tuning-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Explanexploratory Data Analysis</a>
            </span>
        </li>
        <li>
            <span>
                <a href="#Notebook-Configuration" data-toc-modified-id="Notebook-Configuration-2"><span class="toc-item-num">5&nbsp;&nbsp;</span>Feature Importance</a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Packages" data-toc-modified-id="Warning-2.2"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Principal Component Analysis</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Warning" data-toc-modified-id="Warning-2.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Random Forest with Gini Importance</a>
                    </span>
                </li>
              <ul>
                     <span>
                        <a href="#Random-seed" data-toc-modified-id="Random-seed-2.5"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Separating the training, validation and test data</a>
                    </span>
              </ul>
              <ul>
                     <span>
                        <a href="#Random-seed" data-toc-modified-id="Random-seed-2.5"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Ploting top 15 Feature Importance</a>
                    </span>
              </ul>
                </li>
            </ul>
        </li>      
        <li>
            <span>
                <a href="#Generating-the-Submission-File" data-toc-modified-id="Generating-the-Submission-File-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Modeling</a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Creating-the-directory-for-the-submission-file" data-toc-modified-id="Creating-the-directory-for-the-submission-file-6.1">
                            <span class="toc-item-num">6.1&nbsp;&nbsp;</span>Detecting Multi-Colinearity
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Generating-the-submission-file" data-toc-modified-id="Generating-the-submission-file-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Model Selection</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Generating-the-submission-file" data-toc-modified-id="Generating-the-submission-file-6.2"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Defining Models and Parameters</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Generating-the-submission-file" data-toc-modified-id="Generating-the-submission-file-6.2"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Interpretation and Results</a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Generating-the-submission-file" data-toc-modified-id="Generating-the-submission-file-6.2"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Summary</a>
                    </span>
                </li>
            </ul>
        </li>
        <li>
            <span>
                <a href="#Generating-the-Submission-File" data-toc-modified-id="Generating-the-Submission-File-6"><span class="toc-item-num">7&nbsp;&nbsp;</span>Business: Insights and Recommendations</a>
            </span>
            <ul class="toc-item">
            </ul>
        </li>
        <li>
            <span>
                <a href="#Generating-the-Submission-File" data-toc-modified-id="Generating-the-Submission-File-6"><span class="toc-item-num">8&nbsp;&nbsp;</span>Conclusion</a>
            </span>
            <ul class="toc-item">
            </ul>
        </li>
        <li>
            <span>
                <a href="#Generating-the-Submission-File" data-toc-modified-id="Generating-the-Submission-File-6"><span class="toc-item-num">9&nbsp;&nbsp;</span>References</a>
            </span>
            <ul class="toc-item">
            </ul>
        </li>
    </ul>
</div>


<style>
.toc {
  font-family: Avenir;
  font-size: 16px;
}
</style>

## Exordium 

<p style="font-size: 17px"> Are you tired of waiting for your food delivery to arrive while your stomach growls in protest? Well, what if I told you that there's a dataset out there that could help predict delivery times for all your favorite food delivery services? Enter the "Delivery Duration Prediction" dataset.</p>

<p style="font-size: 17px"> This dataset contains information about over 2 Million food deliveries, including the time of day, the distance between the restaurant and customer, the food preparation time, and the actual delivery time. With all this data, it's possible to build a predictive model that could help estimate how long it will take for your next meal to arrive at your doorstep.</p>

<p style="font-size: 17px"> Whether you're a foodie, a data scientist, or just someone who hates waiting for food, this dataset is sure to pique your interest. So why not take a closer look and see what insights we can uncover?</p>

## Notebook Configuration 

### Packages

In [1]:
# Importing all the neccessary packages

import pandas as pd 
import numpy as np
import warnings
import plotly.express as px
import plotly
import plotly.figure_factory as ff
import plotly.graph_objs as go
from plotly.subplots import make_subplots
import plotly.io as pio
import dash as dash
from sklearn.model_selection import train_test_split
from scipy.stats import norm
from utilities import *
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, accuracy_score

# Model Libraries
from xgboost import XGBRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression

### Warning

In [2]:
# Supressing warnings
warnings.filterwarnings('ignore')

#### Global Options

In [3]:
pio.templates.default = "plotly_dark"
pd.set_option('display.float_format', lambda x: '%.2f' % x)

### Random Seed

In [4]:
# Selecting random seed
random_seed = 123

# Setting random seed
np.random.seed(random_seed)

## Data Preprocessing

### Data Description

<table>
    <tr>
        <th>Time features</th>
        <th></th>
    </tr>
    <tr>
        <td>market_id</td>
        <td>A city/region in which DoorDash operates, e.g., Los Angeles, given in the data as an id</td>
    </tr>
    <tr>
        <td>created_at</td>
        <td>Timestamp in UTC when the order was submitted by the consumer to DoorDash. (Note this timestamp is in UTC, but in case you need it, the actual timezone of the region was US/Pacific)</td>
    </tr>
    <tr>
        <td>actual_delivery_time</td>
        <td>Timestamp in UTC when the order was delivered to the consumer</td>
    </tr>
    <tr>
        <th>Store features</th>
        <th></th>
    </tr>
    <tr>
        <td>store_id</td>
        <td>An id representing the restaurant the order was submitted for</td>
    </tr>
    <tr>
        <td>store_primary_category</td>
        <td>Cuisine category of the restaurant, e.g., italian, asian</td>
    </tr>
    <tr>
        <td>order_protocol</td>
        <td>A store can receive orders from DoorDash through many modes. This field represents an id denoting the protocol</td>
    </tr>
    <tr>
        <th>Order features</th>
        <th></th>
    </tr>
    <tr>
        <td>total_items</td>
        <td>Total number of items in the order</td>
    </tr>
    <tr>
        <td>subtotal</td>
        <td>Total value of the order submitted (in cents)</td>
    </tr>
    <tr>
        <td>num_distinct_items</td>
        <td>Number of distinct items included in the order</td>
    </tr>
    <tr>
        <td>min_item_price</td>
        <td>Price of the item with the least cost in the order (in cents)</td>
    </tr>
    <tr>
        <td>max_item_price</td>
        <td>Price of the item with the highest cost in the order (in cents)</td>
    </tr>
    <tr>
        <th>Market features</th>
        <th></th>
    </tr>
    <tr>
        <td>total_onshift_dashers</td>
        <td>Number of available dashers who are within 10 miles of the store at the time of order creation</td>
    </tr>
    <tr>
        <td>total_busy_dashers</td>
        <td>Subset of above total_onshift_dashers who are currently working on an order</td>
    </tr>
    <tr>
        <td>total_outstanding_orders</td>
        <td>Number of orders within 10 miles of this order that are currently being processed</td>
    </tr>
    <tr>
        <th>Predictions from other models</th>
        <th></th>
    </tr>
    <tr>
        <td>estimated_order_place_duration</td>
        <td>Estimated time for the restaurant to receive the order from DoorDash (in seconds)</td>
    </tr>
    <tr>
        <td>estimated_store_to_consumer_driving_duration</td>
        <td>Estimated travel time between store and consumer (in seconds)</td>
    </tr>
</table>


### Loading the data

In [5]:
# Loading the data
df_raw = pd.DataFrame(pd.read_csv('Data/Orders.csv'))

# Making a copy of raw data
df = df_raw.copy()

# Gettin the name of the target
target = 'eta'

#### Looking up Market ID to City Name

<p style= 'color:#ff6666'> Please note that this translation is an approximation and has been compiled from various sources on the internet.</p>

In [6]:
# Laod csv file containing lookup
market_to_city = pd.read_csv('Data/market_to_city.csv', index_col='market_id')['city_name'].to_dict()

# Mapping the Marked ID to City name
df['market_id'] = df['market_id'].map(market_to_city)

In [7]:
df.market_id.unique()

array(['Chicago', 'Phoenix', 'Philadelphia', 'New York', nan, 'Houston',
       'Los Angeles'], dtype=object)

In [8]:
# Printing data demension
display_dimensions(df)

No. of Rows  No. of Columns
0       197428              16

In [9]:
# Glimpse into the data
df.head()

market_id           created_at actual_delivery_time  store_id  \
0       Chicago  2015-02-06 22:24:17  2015-02-06 23:27:16      1845   
1       Phoenix  2015-02-10 21:49:25  2015-02-10 22:56:29      5477   
2  Philadelphia  2015-01-22 20:39:28  2015-01-22 21:09:09      5477   
3  Philadelphia  2015-02-03 21:21:45  2015-02-03 22:13:00      5477   
4  Philadelphia  2015-02-15 02:40:36  2015-02-15 03:20:26      5477   

  store_primary_category  order_protocol  total_items  subtotal  \
0               american            1.00            4      3441   
1                mexican            2.00            1      1900   
2                    NaN            1.00            1      1900   
3                    NaN            1.00            6      6900   
4                    NaN            1.00            3      3900   

   num_distinct_items  min_item_price  max_item_price  total_onshift_dashers  \
0                   4             557            1239                  33.00   
1                   1            1400            1400                   1.00   
2                   1            1900            1900                   1.00   
3                   5             600            1800                   1.00   
4                   3            1100            1600                   6.00   

   total_busy_dashers  total_outstanding_orders  \
0               14.00                     21.00   
1                2.00                      2.00   
2                0.00                      0.00   
3                1.00                      2.00   
4                6.00                      9.00   

   estimated_order_place_duration  \
0                             446   
1                             446   
2                             446   
3                             446   
4                             446   

   estimated_store_to_consumer_driving_duration  
0                                        861.00  
1                                        690.00  
2                                        690.00  
3                                        289.00  
4                                        650.00

In [10]:
df.describe(include='all')

market_id           created_at actual_delivery_time  store_id  \
count     196441               197428               197421 197428.00   
unique         6               180985               178110       NaN   
top      Phoenix  2015-02-11 19:50:43  2015-02-11 20:40:45       NaN   
freq       55058                    6                    5       NaN   
mean         NaN                  NaN                  NaN   3530.51   
std          NaN                  NaN                  NaN   2053.50   
min          NaN                  NaN                  NaN      1.00   
25%          NaN                  NaN                  NaN   1686.00   
50%          NaN                  NaN                  NaN   3592.00   
75%          NaN                  NaN                  NaN   5299.00   
max          NaN                  NaN                  NaN   6987.00   

       store_primary_category  order_protocol  total_items  subtotal  \
count                  192668       196433.00    197428.00 197428.00   
unique                     74             NaN          NaN       NaN   
top                  american             NaN          NaN       NaN   
freq                    19399             NaN          NaN       NaN   
mean                      NaN            2.88         3.20   2682.33   
std                       NaN            1.50         2.67   1823.09   
min                       NaN            1.00         1.00      0.00   
25%                       NaN            1.00         2.00   1400.00   
50%                       NaN            3.00         3.00   2200.00   
75%                       NaN            4.00         4.00   3395.00   
max                       NaN            7.00       411.00  27100.00   

        num_distinct_items  min_item_price  max_item_price  \
count            197428.00       197428.00       197428.00   
unique                 NaN             NaN             NaN   
top                    NaN             NaN             NaN   
freq                   NaN             NaN             NaN   
mean                  2.67          686.22         1159.59   
std                   1.63          522.04          558.41   
min                   1.00          -86.00            0.00   
25%                   1.00          299.00          800.00   
50%                   2.00          595.00         1095.00   
75%                   3.00          949.00         1395.00   
max                  20.00        14700.00        14700.00   

        total_onshift_dashers  total_busy_dashers  total_outstanding_orders  \
count               181166.00           181166.00                 181166.00   
unique                    NaN                 NaN                       NaN   
top                       NaN                 NaN                       NaN   
freq                      NaN                 NaN                       NaN   
mean                    44.81               41.74                     58.05   
std                     34.53               32.15                     52.66   
min                     -4.00               -5.00                     -6.00   
25%                     17.00               15.00                     17.00   
50%                     37.00               34.00                     41.00   
75%                     65.00               62.00                     85.00   
max                    171.00              154.00                    285.00   

        estimated_order_place_duration  \
count                        197428.00   
unique                             NaN   
top                                NaN   
freq                               NaN   
mean                            308.56   
std                              90.14   
min                               0.00   
25%                             251.00   
50%                             251.00   
75%                             446.00   
max                            2715.00   

        estimated_store_to_consumer_driving_duration  
count                  

### Splitting the data
The code below will divide the data into training, validation (80%) and testing (20%).

In [11]:
# Dividing the raw data into training (80%) and testing (20%)
df_train, df_test = train_test_split(df, train_size=0.8, random_state=random_seed)

# Reset the index
df_train, df_test = df_train.reset_index(drop = True), df_test.reset_index(drop = True)

# Dividing the training data into training (80%) and validation (20%)
df_train, df_val = train_test_split(df_train, train_size= 0.8 ,random_state=random_seed)

# Reset the index
df_train, df_val = df_train.reset_index(drop = True), df_val.reset_index(drop= True)

In [12]:
# Printing training data demension
display_dimensions(df_train)

No. of Rows  No. of Columns
0       126353              16

In [13]:
# Printing validation data demension
display_dimensions(df_val)

No. of Rows  No. of Columns
0        31589              16

In [14]:
# Printing test data demension
display_dimensions(df_test)

No. of Rows  No. of Columns
0        39486              16

### Handling identifiers
#### Combining the training, validation and test data

In [15]:
# Combining df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)

In [16]:
#Checking if there is a ID column in the dataframe
df_id = id_checker(df)

df_id.head()

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

### Handling missing data

#### Combining the training, validation and test data

In [17]:
# Combining all three datasets
df = pd.concat([df_train, df_test, df_val], sort= False)

#### Identifying missing values

In [18]:
pd.set_option('display.float_format', lambda x: '%.5f' % x)

# Checking for na values
df_nan = nan_checker(df)

# Printing df_nan
df_nan

var  proportion    dtype
0                         total_onshift_dashers     0.08237  float64
1                            total_busy_dashers     0.08237  float64
2                      total_outstanding_orders     0.08237  float64
3                        store_primary_category     0.02411   object
4                                order_protocol     0.00504  float64
5                                     market_id     0.00500   object
6  estimated_store_to_consumer_driving_duration     0.00266  float64
7                          actual_delivery_time     0.00004   object

In [19]:
# Resetting the precision to 2 decimal
pd.set_option('display.float_format', lambda x: '%.2f' % x)

#### Separating the training, validation and test data

In [20]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [21]:
# Print the dimension of df_train
display_dimensions(df_train)

No. of Rows  No. of Columns
0       126353              16

In [22]:
# Print the dimension of df_val
pd.DataFrame([[df_val.shape[0], df_val.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0   31589         16

In [23]:
# Print the dimension of df_test
pd.DataFrame([[df_test.shape[0], df_test.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0   39486         16

### Imputing missing values
Using the mean of a numerical feature to impute its missing values.

In [24]:
col_to_impute = ['total_onshift_dashers', 'total_busy_dashers', 'total_outstanding_orders']

# The SimpleImputer
si = SimpleImputer(missing_values=np.nan, strategy='mean')

# Imputing the variables with missing values in df_train, df_val and df_test 
df_train[col_to_impute] = si.fit_transform(df_train[col_to_impute])
df_val[col_to_impute] = si.transform(df_val[col_to_impute])
df_test[col_to_impute] = si.transform(df_test[col_to_impute])

### Dropping missing values

In [25]:
df_train.dropna(inplace= True)
df_val.dropna(inplace= True)
df_test.dropna(inplace= True)

### Mapping data types

#### Combining the training, validation and test data

In [26]:
# Combining all three datasets
df = pd.concat([df_train, df_test, df_val], sort= False)

#### Numerical Features

In [27]:
for col in df.columns:
    if df[col].dtype == 'float64':
        df[col] = df[col].astype('int64')

#### TimeStamp Features

In [28]:
# Mapping data time vaiables
datetime_vars = ['created_at','actual_delivery_time']

for col in datetime_vars:
    df[col] = pd.to_datetime(df[col])

#### Categorical Features

In [29]:
# Mapping categorical vvariable
col_to_cat = ['market_id','store_primary_category','order_protocol']
for col in col_to_cat: 
    df[col] = pd.Categorical(df[col])


#### Improving Readability

In [30]:
# Convert subtotal from cents to dollar
df['subtotal'] = df['subtotal'] / 100
# Convert subtotal from cents to dollar
df['min_item_price'] = df['min_item_price'] / 100
# Convert subtotal from cents to dollar
df['max_item_price'] = df['max_item_price'] / 100


#### Separating the training, validation and test data

In [31]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [32]:
# Print the dimension of df_train
display_dimensions(df_train)

No. of Rows  No. of Columns
0       122560              16

In [33]:
# Print the dimension of df_val
pd.DataFrame([[df_val.shape[0], df_val.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0   30633         16

In [34]:
# Print the dimension of df_test
pd.DataFrame([[df_test.shape[0], df_test.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0   38214         16

### Removing unnecessary features

#### Combining the training, validation and test data

In [35]:
# Combine df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)

In [36]:
col_to_remove = ['store_id']

In [37]:
# Remove the uncommon features from the training data
df.drop(columns=col_to_remove, inplace= True)

In [38]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [39]:
# Print the dimension of df_train
display_dimensions(df_train)

No. of Rows  No. of Columns
0       122560              15

In [40]:
# Print the dimension of df_test
display_dimensions(df_test)

No. of Rows  No. of Columns
0        38214              15

In [41]:
# Print the dimension of df_val
display_dimensions(df_val)

No. of Rows  No. of Columns
0        30633              15

### Creating a Duplicate Dataset for EDA and Modeling

#### Combining the training, validation and test data

In [42]:
# Combine df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)

In [43]:
# Making a copy for EDA before encoding
df_eda = df.copy()

### Encoding the data

#### Identifying categorical variables

In [44]:
# Print the unique data type of variables in df
pd.DataFrame(df.dtypes.unique(), columns=['dtype'])

dtype
0        category
1  datetime64[ns]
2        category
3        category
4           int64
5         float64

In [45]:
# Calling cat_var_checker on df
df_cat = cat_var_checker(df, dtype='category')

# Print the dataframe
df_cat

var  nunique
0  store_primary_category       74
1          order_protocol        7
2               market_id        6

#### Encoding categorical features

In [46]:
# One-hot-encode the categorical features in the combined data
df = pd.get_dummies(df, columns= np.setdiff1d(df_cat['var'], [target]))

# Print the first 5 rows of df
df.head()

created_at actual_delivery_time  total_items  subtotal  \
0 2015-02-06 19:54:18  2015-02-06 20:31:38            2     19.28   
1 2015-02-13 19:05:05  2015-02-13 19:39:48            4     13.90   
2 2015-02-07 01:23:23  2015-02-07 01:56:22            2     14.90   
3 2015-01-25 01:40:48  2015-01-25 02:40:15            7     72.45   
4 2015-02-10 01:54:29  2015-02-10 02:46:44            1     11.95   

   num_distinct_items  min_item_price  max_item_price  total_onshift_dashers  \
0                   2            8.99           10.29                      7   
1                   3            2.45            3.45                     62   
2                   2            6.95            7.95                    105   
3                   7            4.75           12.65                     40   
4                   1           11.95           11.95                     10   

   total_busy_dashers  total_outstanding_orders  ...  \
0                   7                         9  ...   
1                  58                       119  ...   
2                  52                        78  ...   
3                  40                        48  ...   
4                   8                         7  ...   

   store_primary_category_southern  store_primary_category_spanish  \
0                                0                               0   
1                                0                               0   
2                                0                               0   
3                                0                               0   
4                                0                               0   

   store_primary_category_steak  store_primary_category_sushi  \
0                             0                             0   
1                             0                             0   
2                             0                             0   
3                             0                             0   
4                             0                             0   

   store_primary_category_tapas  store_primary_category_thai  \
0                             0                            0   
1                             0                            0   
2                             0                            0   
3                             0                            0   
4                             0                            0   

   store_primary_category_turkish  store_primary_category_vegan  \
0                               0                             0   
1                               0                             0   
2                               0                             0   
3                               0                             0   
4                               0                             0   

   store_primary_category_vegetarian  store_primary_category_vietnamese  
0                                  0                                  0  
1                                  0                                  0  
2                                  0                                  0  
3                                  0                                  0  
4                                  0                                  0  

[5 rows x 99 columns]

#### Separating the training, validation and test data

In [47]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [48]:
# Print the dimension of df_train
display_dimensions(df_train)

No. of Rows  No. of Columns
0       122560              99

In [49]:
# Print the dimension of df_test
display_dimensions(df_test)

No. of Rows  No. of Columns
0        38214              99

In [50]:
# Print the dimension of df_val
display_dimensions(df_val)

No. of Rows  No. of Columns
0        30633              99

### Handling date and time variable

#### Transforming date time variables


In [51]:
# Getting the date time variables
datetime_vars = ['created_at','actual_delivery_time']

In [52]:
# Calling the datetime transformer on the df_train
df_train = datetime_transformer(df_train, datetime_vars)

# Printing the first 5 rows
df_train.head()

created_at actual_delivery_time  total_items  subtotal  \
0 2015-02-06 19:54:18  2015-02-06 20:31:38            2     19.28   
1 2015-02-13 19:05:05  2015-02-13 19:39:48            4     13.90   
2 2015-02-07 01:23:23  2015-02-07 01:56:22            2     14.90   
3 2015-01-25 01:40:48  2015-01-25 02:40:15            7     72.45   
4 2015-02-10 01:54:29  2015-02-10 02:46:44            1     11.95   

   num_distinct_items  min_item_price  max_item_price  total_onshift_dashers  \
0                   2            8.99           10.29                      7   
1                   3            2.45            3.45                     62   
2                   2            6.95            7.95                    105   
3                   7            4.75           12.65                     40   
4                   1           11.95           11.95                     10   

   total_busy_dashers  total_outstanding_orders  ...  created_at_day  \
0                   7                         9  ...               6   
1                  58                       119  ...              13   
2                  52                        78  ...               7   
3                  40                        48  ...              25   
4                   8                         7  ...              10   

   created_at_hour  created_at_minute  created_at_second  \
0               19                 54                 18   
1               19                  5                  5   
2                1                 23                 23   
3                1                 40                 48   
4                1                 54                 29   

   actual_delivery_time_year  actual_delivery_time_month  \
0                       2015                           2   
1                       2015                           2   
2                       2015                           2   
3                       2015                           1   
4                       2015                           2   

   actual_delivery_time_day  actual_delivery_time_hour  \
0                         6                         20   
1                        13                         19   
2                         7                          1   
3                        25                          2   
4                        10                          2   

   actual_delivery_time_minute  actual_delivery_time_second  
0                           31                           38  
1                           39                           48  
2                           56                           22  
3                           40                           15  
4                           46                           44  

[5 rows x 111 columns]

In [53]:
# Calling the datetime transformer on the df_val
df_val = datetime_transformer(df_val, datetime_vars)

# Printing the first 5 rows
df_val.head()

created_at actual_delivery_time  total_items  subtotal  \
31589 2015-02-11 02:44:42  2015-02-11 03:32:41            4     36.50   
31590 2015-02-13 03:50:03  2015-02-13 04:33:30            2     22.48   
31591 2015-01-23 03:49:50  2015-01-23 04:33:56            4     33.80   
31592 2015-02-03 20:23:15  2015-02-03 20:54:28            2     17.98   
31593 2015-01-23 00:25:46  2015-01-23 01:45:20           30    116.25   

       num_distinct_items  min_item_price  max_item_price  \
31589                   4            3.00           11.00   
31590                   2            2.49           18.99   
31591                   4            3.95           18.95   
31592                   2            7.99            9.99   
31593                   2            3.75            4.00   

       total_onshift_dashers  total_busy_dashers  total_outstanding_orders  \
31589                     92                  86                       125   
31590                     32                  18                        21   
31591                     35                  32                        44   
31592                     45                  46                        60   
31593                     32                  27                        30   

       ...  created_at_day  created_at_hour  created_at_minute  \
31589  ...              11                2                 44   
31590  ...              13                3                 50   
31591  ...              23                3                 49   
31592  ...               3               20                 23   
31593  ...              23                0                 25   

       created_at_second  actual_delivery_time_year  \
31589                 42                       2015   
31590                  3                       2015   
31591                 50                       2015   
31592                 15                       2015   
31593                 46                       2015   

       actual_delivery_time_month  actual_delivery_time_day  \
31589                           2                        11   
31590                           2                        13   
31591                           1                        23   
31592                           2                         3   
31593                           1                        23   

       actual_delivery_time_hour  actual_delivery_time_minute  \
31589                          3                           32   
31590                          4                           33   
31591                          4                           33   
31592                         20                           54   
31593                          1                           45   

       actual_delivery_time_second  
31589                           41  
31590                           30  
31591                           56  
31592                           28  
31593                           20  

[5 rows x 111 columns]

In [54]:
# Calling the datetime transformer on the df_test
df_test = datetime_transformer(df_test, datetime_vars)

# Printing the first 5 rows
df_test.head()

created_at actual_delivery_time  total_items  subtotal  \
23770 2015-01-24 02:41:10  2015-01-24 03:26:50            4     42.77   
23771 2015-01-24 17:52:07  2015-01-24 18:42:49            5     26.50   
23772 2015-02-12 21:52:21  2015-02-12 22:29:47            2     20.28   
23773 2015-02-18 01:46:14  2015-02-18 02:15:20            1     11.95   
23774 2015-01-26 01:41:26  2015-01-26 02:10:45            1      3.50   

       num_distinct_items  min_item_price  max_item_price  \
23770                   4            5.99           14.33   
23771                   4            2.50           13.00   
23772                   2            8.79           11.49   
23773                   1           11.95           11.95   
23774                   1            3.50            3.50   

       total_onshift_dashers  total_busy_dashers  total_outstanding_orders  \
23770                     13                  19                        15   
23771                     10                  33                        15   
23772                     28                  21                        21   
23773                     73                  71                       114   
23774                     44                  41                        58   

       ...  created_at_day  created_at_hour  created_at_minute  \
23770  ...              24                2                 41   
23771  ...              24               17                 52   
23772  ...              12               21                 52   
23773  ...              18                1                 46   
23774  ...              26                1                 41   

       created_at_second  actual_delivery_time_year  \
23770                 10                       2015   
23771                  7                       2015   
23772                 21                       2015   
23773                 14                       2015   
23774                 26                       2015   

       actual_delivery_time_month  actual_delivery_time_day  \
23770                           1                        24   
23771                           1                        24   
23772                           2                        12   
23773                           2                        18   
23774                           1                        26   

       actual_delivery_time_hour  actual_delivery_time_minute  \
23770                          3                           26   
23771                         18                           42   
23772                         22                           29   
23773                          2                           15   
23774                          2                           10   

       actual_delivery_time_second  
23770                           50  
23771                           49  
23772                           47  
23773                           20  
23774                           45  

[5 rows x 111 columns]

### Feature Engineering

<table border="1" cellpadding="5" cellspacing="0">
  <thead>
    <tr>
      <th>Feature Name</th>
      <th>Original Feature 1</th>
      <th>Original Feature 2</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Target Variable (ETA)</td>
      <td>created_at</td>
      <td>actual_delivery_time</td>
      <td>The estimated time of arrival for the order</td>
    </tr>
    <tr>
      <td>Busy Rider Ratio (%)</td>
      <td>total_onshift_dashers</td>
      <td>total_busy_dashers</td>
      <td>The percentage of on-shift dashers who are currently occupied with an order</td>
    </tr>
    <tr>
      <td>Non-Preparation Duration</td>
      <td>estimated_order_place_duration</td>
      <td>estimated_store_to_consumer_driving_duration</td>
      <td>The duration of the delivery process, excluding the time spent at the restaurant for preparation</td>
    </tr>
    <tr>
      <td>Day of the week</td>
      <td>created_at</td>
      <td>-</td>
      <td>The day of the week when the order was placed</td>
    </tr>
    <tr>
      <td>Average Order Price</td>
      <td>subtotal</td>
      <td>total_items</td>
      <td>The average price of items in the order</td>
    </tr>
    <tr>
      <td>Time of the day</td>
      <td>created_at</td>
      <td>-</td>
      <td>The time of day when the order was placed</td>
    </tr>
    <tr>
      <td>Delivery Duration</td>
      <td>eta</td>
      <td>-</td>
      <td>The total duration of the delivery process is categorized as <15 mins, 15 - 30 mins, 30 - 45 mins, 45 - 60 mins, and >60 mins</td>
    </tr>
    <tr>
      <td>Trip Nature</td>
      <td>eta</td>
      <td>-</td>
      <td>The nature of the delivery trip, categorized as Fast, Moderate, Slow & Very Slow</td>
    </tr>
    <tr>
      <td>Delivery distance</td>
      <td>estimated_store_to_consumer_driving_duration</td>
      <td>-</td>
      <td>The estimated distance between the store and the consumer</td>
    </tr>
    <tr>
      <td>Order Density</td>
      <td>total_outstanding_orders</td>
      <td>total_busy_dashers, total_onshift_dashers</td>
      <td>The number of orders within 10 miles of the order being processed</td>
    </tr>
  </tbody>
</table>


#### Combining the training, validation and test data

In [55]:
# Combine df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)

#### Target Variable (ETA)

In [56]:
# Calculating the time diff from order created to order recieved
df['eta'] =  (df['actual_delivery_time'] - df['created_at']).dt.total_seconds().astype(int)
df_eda['eta'] =  (df_eda['actual_delivery_time'] - df_eda['created_at']).dt.total_seconds().astype(int)

#### Busy Rider Ratio (%)

In [57]:
# Calculating the available riders as a ration of total riders 
df['busy_rider_ratio'] = df['total_busy_dashers'] / df['total_onshift_dashers'].apply(lambda x: x if x != 0 else 1)
df_eda['busy_rider_ratio'] = df_eda['total_busy_dashers'] / df_eda['total_onshift_dashers'].apply(lambda x: x if x != 0 else 1)

# Replacing values where total_onshift_dashers are zero which leads to busy_rider_ratio to inf by mean

df_eda['busy_rider_ratio'] = df_eda['busy_rider_ratio'].replace([np.inf, -np.inf], df_eda['busy_rider_ratio'].mean())
df['busy_rider_ratio'] = df['busy_rider_ratio'].replace([np.inf, -np.inf], df['busy_rider_ratio'].mean())

#### Non-Preparation Duration

In [58]:
'''The delivery duration can be broken down to 
1. time for restuant to recieve the order (estimated_order_place_duration)
2. time for preparation
3. time of travel from store to customer''' 

df['non_prep_duration'] =  df['estimated_order_place_duration'] + df['estimated_store_to_consumer_driving_duration']
df_eda['non_prep_duration'] =  df_eda['estimated_order_place_duration'] + df_eda['estimated_store_to_consumer_driving_duration']

#### Day of the week

In [59]:
# Adding the day of the week
df['day_of_week'] = df['created_at'].dt.day_name().astype('category')
df_eda['day_of_week'] = df_eda['created_at'].dt.day_name().astype('category')

#### Average Order Price

In [60]:
# Calulating avg price per order

df['avg_price'] = df['subtotal'] / df['total_items']
df_eda['avg_price'] = df_eda['subtotal'] / df_eda['total_items']

# Replacing values where total_items are zero which leads to avg_price to inf by mean
df_eda['avg_price'] = df_eda['avg_price'].replace([np.inf, -np.inf], df_eda['avg_price'].mean())
df['avg_price'] = df['avg_price'].replace([np.inf, -np.inf], df['avg_price'].mean())


#### Time of the day 

In [61]:
# Cut data into day time sections
df_eda['time_of_day'] = pd.cut(df_eda['created_at'].dt.hour, bins=[0, 6, 12, 18, 24], labels=['Night', 'Morning', 'Afternoon', 'Evening'], include_lowest=True).astype('category')


#### Delivery Duration

In [62]:
# Defining buckets for eta
delivery_time_bins = [0,15,30,45,60,float('inf')]
# Defining labels
delivery_time_labels = ['<15 mins', '15 - 30 mins', '30 - 45 mins', '45 - 60 mins', '> 60 mins']

# Adding delivery duration
df_eda['delivery_duration'] = pd.cut(df['eta']/ 60,
                                        bins = delivery_time_bins, 
                                        labels= delivery_time_labels,
                                        include_lowest=True).astype('category')


#### Trip Nature

In [63]:
# Defining buckets for eta
delivery_nature_bins = [0,20,40,60,float('inf')]
# Defining labels
delivery_nature_labels = ['Fast', 'Moderate', 'Slow', 'Very Slow']

# Adding delivery duration
df_eda['delivery_nature'] = pd.cut(df_eda['eta']/ 60,
        bins = delivery_nature_bins, 
        labels= delivery_nature_labels,
        include_lowest=True).astype('category')

#### Delivery distance

In [64]:
# We can estimate the distance assuming constant speed 17 miles/hour
df['delivery_distance'] = (df['estimated_store_to_consumer_driving_duration'] / 3600) * 17
df_eda['delivery_distance'] = (df_eda['estimated_store_to_consumer_driving_duration'] / 3600) * 17

#### Order Density

In [65]:
# Order density means on average, each busy dasher has how many outstanding orders waiting to be delivered.

df['order_density'] = df['total_outstanding_orders'] / (df['total_busy_dashers'] + df['total_onshift_dashers'])
df_eda['order_density'] = df_eda['total_outstanding_orders'] / (df_eda['total_busy_dashers'] + df_eda['total_onshift_dashers'])

# Replacing values where total_busy_dashers & total_onshift_dashers are zero which leads to order_density to inf by mean
mean_order_density = df['order_density'].replace([np.inf, -np.inf], np.nan).mean(skipna=True)

df_eda['order_density'] = df_eda['order_density'].replace([np.inf, -np.inf, np.nan], mean_order_density.mean())
df['order_density'] = df['order_density'].replace([np.inf, -np.inf, np.nan], mean_order_density.mean())

### Outlier Detection and Treatment

#### Outliers Detection

In [66]:
outlier_detection(df_eda)

Column  Outlier Count  Percentage
1                                       subtotal           3673        1.92
4                                 max_item_price           2989        1.56
13                                     avg_price           2949        1.54
2                             num_distinct_items           2901        1.52
3                                 min_item_price           2700        1.41
0                                    total_items           2479        1.30
7                       total_outstanding_orders           1991        1.04
15                                 order_density           1691        0.88
11                              busy_rider_ratio           1173        0.61
5                          total_onshift_dashers            763        0.40
6                             total_busy_dashers            510        0.27
12                             non_prep_duration            314        0.16
9   estimated_store_to_consumer_driving_duration            263        0.14
14                             delivery_distance            263        0.14
8                 estimated_order_place_duration            130        0.07
10                                           eta              3        0.00

#### Outlier Treatment

#### Outlier Rules for Dataset Ranges

<table>
  <thead>
    <tr>
      <th>Feature</th>
      <th>From</th>
      <th>To</th>
      <th>Comments</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>total_items</td>
      <td>0</td>
      <td>15</td>
      <td>Let's keep it small like a family dinner</td>
    </tr>
    <tr>
      <td>subtotal</td>
      <td>0</td>
      <td>$90</td>
      <td>No more extra guac please</td>
    </tr>
    <tr>
      <td>num_distinct_items</td>
      <td>0</td>
      <td>15</td>
      <td>Can't handle too many choices</td>
    </tr>
    <tr>
      <td>min_item_price</td>
      <td>0</td>
      <td>Up to max observed</td>
      <td>No more discounts for negative prices</td>
    </tr>
    <tr>
      <td>total_onshift_dashers</td>
      <td>0</td>
      <td>Up to max observed</td>
      <td>Let's keep it positive for our riders</td>
    </tr>
    <tr>
      <td>total_busy_dashers</td>
      <td>0</td>
      <td>Up to max observed</td>
      <td>Can't have negative busy dashers</td>
    </tr>
    <tr>
      <td>total_outstanding_orders</td>
      <td>0</td>
      <td>Up to max observed</td>
      <td>No negative orders, we are not in debt here</td>
    </tr>
    <tr>
      <td>eta</td>
      <td>0</td>
      <td>1.5 hours</td>
      <td>No one should wait for their food for more than 1.5 hrs</td>
    </tr>
    <tr>
      <td>busy_rider_ratio</td>
      <td>0</td>
      <td>Up to max observed</td>
      <td>Our riders are always busy, in a good way</td>
    </tr>
    <tr>
      <td>non_prep_duration</td>
      <td>0</td>
      <td>1.5 hours</td>
      <td>Let's prepare the food and deliver it ASAP</td>
    </tr>
  </tbody>
</table>

#### Defining Range

In [67]:
# Restrict the range for total_items
df_eda = df_eda[df_eda['total_items'] <= 15]
df = df[df['total_items'] <= 15]

# Restrict the range for subtotal
df_eda = df_eda[df_eda['subtotal'].between(0, 90)]
df = df[df['subtotal'].between(0, 90)]

# Restrict the range for num_distinct_items
df_eda = df_eda[df_eda['num_distinct_items'] <= 15]
df = df[df['num_distinct_items'] <= 15]

# Restrict the range for min_item_price
df_eda = df_eda[df_eda['min_item_price'] >= 0]
df = df[df['min_item_price'] >= 0]

# Restrict the range for total_onshift_dashers
df_eda = df_eda[df_eda['total_onshift_dashers'] >= 0]
df = df[df['total_onshift_dashers'] >= 0]

# Restrict the range for total_busy_dashers
df_eda = df_eda[df_eda['total_busy_dashers'] >= 0]
df = df[df['total_busy_dashers'] >= 0]

# Restrict the range for total_outstanding_orders
df_eda = df_eda[df_eda['total_outstanding_orders'] >= 0]
df = df[df['total_outstanding_orders'] >= 0]

# Restrict the range for eta
df_eda = df_eda[df_eda['eta'].between(0, 5400)]
df = df[df['eta'].between(0, 5400)]

# Restrict the range for busy_rider_ratio
df_eda = df_eda[df_eda['busy_rider_ratio'] >= 0]
df = df[df['busy_rider_ratio'] >= 0]

# Restrict the range for non_prep_duration
df_eda = df_eda[df_eda['non_prep_duration'].between(0, 5400)]
df = df[df['non_prep_duration'].between(0, 5400)]

# Restrict the range for order_density
df = df[df['order_density'] >= 0]


### Normality Test

In [68]:
import pandas as pd
from scipy.stats import shapiro, normaltest, kstest

# Assuming df_eda is your dataframe
feature_names = df_eda.select_dtypes(['int64', 'float64']).columns

# Initialize lists to store test results
shapiro_results = []
dagostino_results = []
ks_results = []

# Iterate through features and perform tests
for feature in feature_names:
    shapiro_stat, shapiro_p = shapiro(df_eda[feature])
    dagostino_stat, dagostino_p = normaltest(df_eda[feature])
    ks_stat, ks_p = kstest(df_eda[feature], 'norm')
    
    shapiro_results.append(shapiro_p)
    dagostino_results.append(dagostino_p)
    ks_results.append(ks_p)

# Create a results dataframe
results_df = pd.DataFrame({'Feature': feature_names,
                           'Shapiro_p': shapiro_results,
                           'Dagostino_p': dagostino_results,
                           'KS_p': ks_results})

results_df


Feature  Shapiro_p  Dagostino_p  KS_p
0                                    total_items       0.00         0.00  0.00
1                                       subtotal       0.00         0.00  0.00
2                             num_distinct_items       0.00         0.00  0.00
3                                 min_item_price       0.00         0.00  0.00
4                                 max_item_price       0.00         0.00  0.00
5                          total_onshift_dashers       0.00         0.00  0.00
6                             total_busy_dashers       0.00         0.00  0.00
7                       total_outstanding_orders       0.00         0.00  0.00
8                 estimated_order_place_duration       0.00         0.00  0.00
9   estimated_store_to_consumer_driving_duration       0.00         0.00  0.00
10                                           eta       0.00         0.00  0.00
11                              busy_rider_ratio       0.00         0.00  0.00
12                             non_prep_duration       0.00         0.00  0.00
13                                     avg_price       0.00         0.00  0.00
14                             delivery_distance       0.00         0.00  0.00
15                                 order_density       0.00         0.00  0.00

### Scaling the data

#### Standardization

In [69]:
# Scale the numerical columns
num_cols = df.select_dtypes(include=['float64', 'int64']).columns
scaler = StandardScaler()
df[num_cols] = scaler.fit_transform(df[num_cols])

## Explanexploratory Data Analysis

#### [Explore the DoorDash Data with an Interactive Dashboard - Click Here!](https://dashapp-3h7gsvk4ca-nn.a.run.app/)

## Feature Importance

In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 183063 entries, 0 to 31588
Columns: 118 entries, created_at to order_density
dtypes: category(1), datetime64[ns](2), float64(28), uint8(87)
memory usage: 58.7 MB


### Principal Component Analysis

In [71]:
num_cols = df.select_dtypes(include=['float64', 'int64','uint8']).columns

# Run a PCA on the data
pca = PCA(n_components=0.95)
pca.fit(df[num_cols])

# Calculate the number of components needed to explain 95% of the variance
num_components = len(pca.explained_variance_ratio_)

# Create a line plot of the cumulative explained variance ratio
fig = px.line(x=range(1, num_components+1),
              y=np.cumsum(pca.explained_variance_ratio_),
              labels={'x': 'Number of Components', 'y': 'Cumulative Explained Variance Ratio'},
              color_discrete_sequence=['#636EFA'])

# Add horizontal and vertical lines at 95%
fig.add_hline(y=0.95, line_width=3, line_dash='dash', line_color='red', annotation_text='95% Explained Variance Ratio')
fig.add_vline(x=num_components, line_width=3, line_dash='dash', line_color='red')

# Update the layout
fig.update_layout(title={'x':0.5, 'xanchor': 'center', 'text': 'Cumulative Explained Variance Ratio by Number of Components'},
                xaxis=dict(title='Number of Components',
                             #tickmode='array',
                             #tickvals=[0, 5, 10, 15, 20, num_components, 25]
                             ))

# Show the plot
fig.show()


### Random Forest with Gini Importance

#### Separating the training, validation and test data

In [72]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [73]:
# Perform one-hot encoding on the 'day_of_week' column
enc = OneHotEncoder(sparse=False, handle_unknown='ignore')
encoded_day_of_week = enc.fit_transform(df[['day_of_week']])
encoded_day_of_week_df = pd.DataFrame(encoded_day_of_week, columns=enc.get_feature_names_out(['day_of_week']))

# Reset the index of the DataFrames
df = df.reset_index(drop=True)
encoded_day_of_week_df = encoded_day_of_week_df.reset_index(drop=True)

# Create a new DataFrame without the 'created_at', 'actual_delivery_time', and 'day_of_week' columns,
# and with the one-hot encoded 'day_of_week' columns
df_processed = pd.concat([df.drop(['created_at', 'day_of_week'] + [col for col in df.columns if col.startswith('actual_delivery_time')], axis=1), encoded_day_of_week_df], axis=1)

# Prepare the dataset for training
X = df_processed.drop('eta', axis=1)
y = df_processed['eta']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fit the Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

# Calculate the Gini Importance of the features
importances = rf.feature_importances_
feature_importances = pd.Series(importances, index=X.columns).sort_values(ascending=False)




#### Ploting top 15 Feature Importance

In [74]:
# Plot the feature importances as a bar plot top 10 features
fig = px.bar(x=feature_importances[:15].index,
                y=feature_importances[:15].values,
                labels={'x': 'Features', 'y': 'Importance'},
                color=feature_importances[:15].values)
fig.update_layout(title={'x': 0.5, 'xanchor': 'center', 'text': 'Top 15 Feature Importances'})

fig.show()

## Modeling

### Detecting Multi-Colinearity

In [75]:
def get_top_correlated_features(df, target_column, threshold=None, num_top_features=None):
    """
    Get the top correlated features with the target variable.

    :param df: DataFrame containing the features and target variable
    :param target_column: Name of the target column
    :param threshold: Optional correlation threshold to filter features (default: None)
    :param num_top_features: Optional number of top features to display (default: None)
    :return: DataFrame with the top correlated features and their correlation coefficients
    """

    # Calculate the correlation matrix
    corr_matrix = df.corr()

    # Get the correlation coefficients for the target variable
    target_corr = corr_matrix[target_column].drop(target_column)

    # If a threshold is specified, filter the features based on the threshold
    if threshold is not None:
        target_corr = target_corr[abs(target_corr) >= threshold]

    # Sort the features by their absolute correlation coefficient in descending order
    top_correlated_features = target_corr.sort_values(key=abs, ascending=False).reset_index()

    # If num_top_features is specified, return only the specified number of top features
    if num_top_features is not None:
        top_correlated_features = top_correlated_features.head(num_top_features)

    # Rename the columns in the output DataFrame
    top_correlated_features.columns = ['Feature', 'Correlation']

    return top_correlated_features

get_top_correlated_features(df = df_processed, target_column = 'eta', num_top_features = 10)

Feature  Correlation
0                                 order_density         0.35
1                             non_prep_duration         0.28
2                             delivery_distance         0.26
3  estimated_store_to_consumer_driving_duration         0.26
4                      total_outstanding_orders         0.22
5                                      subtotal         0.21
6                               created_at_hour        -0.18
7                            num_distinct_items         0.15
8                                max_item_price         0.14
9                            total_busy_dashers         0.13

In [76]:
def get_redundant_features(df, correlation_threshold=0.8):
    """
    Get redundant feature pairs to detect multicollinearity.

    :param df: DataFrame containing the features
    :param correlation_threshold: Correlation coefficient threshold to consider a pair redundant (default: 0.8)
    :return: DataFrame with redundant feature pairs and their correlation coefficients
    """

    # Calculate the correlation matrix
    corr_matrix = df.corr()

    # Find feature pairs with a correlation coefficient above the threshold
    redundant_pairs = []
    for i in range(len(corr_matrix.columns)):
        for j in range(i+1, len(corr_matrix.columns)):
            if abs(corr_matrix.iloc[i, j]) >= correlation_threshold:
                redundant_pairs.append((corr_matrix.columns[i], corr_matrix.columns[j], corr_matrix.iloc[i, j]))

    # Create a DataFrame with the redundant feature pairs and their correlation coefficients
    redundant_features = pd.DataFrame(redundant_pairs, columns=['Feature 1', 'Feature 2', 'Correlation'])

    # Sort the DataFrame by the absolute correlation coefficient in descending order
    redundant_features = redundant_features.sort_values(by='Correlation', key=abs, ascending=False).reset_index(drop=True)

    return redundant_features

get_redundant_features(df= df_processed, correlation_threshold=0.45)


Feature 1                    Feature 2  \
0   estimated_store_to_consumer_driving_duration            delivery_distance   
1                          total_onshift_dashers           total_busy_dashers   
2                          total_onshift_dashers     total_outstanding_orders   
3                             total_busy_dashers     total_outstanding_orders   
4   estimated_store_to_consumer_driving_duration            non_prep_duration   
5                              non_prep_duration            delivery_distance   
6                 estimated_order_place_duration             order_protocol_1   
7                               created_at_month               created_at_day   
8                                    total_items           num_distinct_items   
9                                 min_item_price                    avg_price   
10                                max_item_price                    avg_price   
11                                      subtotal           num_distinct_items   
12                                   total_items                     subtotal   
13                                min_item_price               max_item_price   
14                                      subtotal               max_item_price   
15                      total_outstanding_orders                order_density   
16                              order_protocol_4  store_primary_category_fast   
17                                   total_items               min_item_price   
18                            num_distinct_items               min_item_price   

    Correlation  
0          1.00  
1          0.94  
2          0.94  
3          0.93  
4          0.92  
5          0.92  
6          0.90  
7         -0.88  
8          0.87  
9          0.86  
10         0.77  
11         0.65  
12         0.62  
13         0.55  
14         0.52  
15         0.52  
16         0.50  
17        -0.47  
18        -0.47

### Model Selection

In [77]:

def run_model(model, model_name, X_train, y_train, X_test, y_test, verbose=True):
    """
    Train and evaluate a model.

    :param model: Model to train and evaluate
    :param model_name: Name of the model
    :param X_train: Training features
    :param y_train: Training target variable
    :param X_test: Test features
    :param y_test: Test target variable
    :param verbose: Whether to print the results (default: True)
    :return: Trained model
    """

    # Train the model
    model.fit(X_train, y_train)

    # Make predictions on the test set
    y_pred = model.predict(X_test)

    # Calculate the metrics
    mae = mean_absolute_error(y_test, y_pred)
    rmse = mean_squared_error(y_test, y_pred, squared=False)
    r2 = r2_score(y_test, y_pred)
    
    if verbose:
        # Print the metrics
        print(f'Model: {model_name}')
        print(f'Mean Absolute Error: {mae:.2f}')
        print(f'Root Mean Squared Error: {rmse:.2f}')
        print(f'R^2 Score: {r2:.2f}')

    # Create a dataframe to store the results
    model_results = pd.DataFrame(data=[[model_name, mae, rmse, r2]],
                                  columns=['Model', 'Mean Absolute Error', 'Root Mean Squared Error', 'R^2 Score'])

    return model, model_results



#### Defining Models and Parameters

* **Models**

    1. XGBoost Regression 
    2. Multi-Layer Perceptron Regressor 
    3. Decision Tree Regression 
    4. Support Vector Regression 
    5. K-Nearest Neighbors Regression 
    6. Linear Regression 

* **Feature Set (As per Gini's Importance)**

    1. Full
    2. 40 Feature Dataset
    3. 20 Feature Dataset
    4. 10 Feature Dataset

In [78]:
models = {'XGBoost Regression': XGBRegressor(random_state=42),
            'Multi-Layer Perceptron Regressor': MLPRegressor(random_state=42),
            'Decision Tree Regression': DecisionTreeRegressor(random_state=42),
            'Support Vector Regression': SVR(),
            'K-Nearest Neighbors Regression': KNeighborsRegressor(),
            'Linear Regression': LinearRegression()}

datasets = {'Top 40 Features': X_train[feature_importances[:40].index].assign(eta=y_train),
            'Top 20 Features': X_train[feature_importances[:20].index].assign(eta=y_train),
            'Top 10 Features': X_train[feature_importances[:10].index].assign(eta=y_train),
            'All Features': X_train.assign(eta=y_train)}

In [80]:
from joblib import Parallel, delayed

def process_model(dataset_name, dataset, model_name, model):
    X = dataset.drop('eta', axis=1)
    y = dataset['eta']
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    _, model_results = run_model(model, model_name, X_train, y_train, X_test, y_test, verbose=False)
    model_results['Dataset'] = dataset_name 
    return model_results

# Parallelize model training and evaluation
num_cores = -1  
results_list = Parallel(n_jobs=num_cores)(
    delayed(process_model)(dataset_name, dataset, model_name, model)
    for dataset_name, dataset in datasets.items()
    for model_name, model in models.items()
)

# Combine the results into a single DataFrame
results = pd.concat(results_list, axis=0, ignore_index=True)

# reset the index of the results dataframe
results = results.reset_index(drop=True)

results


/Users/chiraglakhanpal/PycharmProjects/pythonProject/venv/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Model  Mean Absolute Error  \
0                 XGBoost Regression                 0.63   
1   Multi-Layer Perceptron Regressor                 0.65   
2           Decision Tree Regression                 0.92   
3          Support Vector Regression                 0.64   
4     K-Nearest Neighbors Regression                 0.71   
5                  Linear Regression                 0.67   
6                 XGBoost Regression                 0.64   
7   Multi-Layer Perceptron Regressor                 0.65   
8           Decision Tree Regression                 0.93   
9          Support Vector Regression                 0.65   
10    K-Nearest Neighbors Regression                 0.73   
11                 Linear Regression                 0.68   
12                XGBoost Regression                 0.66   
13  Multi-Layer Perceptron Regressor                 0.66   
14          Decision Tree Regression                 0.95   
15         Support Vector Regression                 0.66   
16    K-Nearest Neighbors Regression                 0.73   
17                 Linear Regression                 0.69   
18                XGBoost Regression                 0.62   
19  Multi-Layer Perceptron Regressor                 0.64   
20          Decision Tree Regression                 0.91   
21         Support Vector Regression                 0.63   
22    K-Nearest Neighbors Regression                 0.70   
23                 Linear Regression                 0.67   

    Root Mean Squared Error  R^2 Score          Dataset  
0                      0.80       0.36  Top 40 Features  
1                      0.82       0.32  Top 40 Features  
2                      1.18      -0.39  Top 40 Features  
3                      0.83       0.32  Top 40 Features  
4                      0.90       0.19  Top 40 Features  
5                      0.85       0.28  Top 40 Features  
6                      0.81       0.35  Top 20 Features  
7                      0.83       0.32  Top 20 Features  
8                      1.19      -0.41  Top 20 Features  
9                      0.85       0.28  Top 20 Features  
10                     0.93       0.15  Top 20 Features  
11                     0.85       0.27  Top 20 Features  
12                     0.84       0.30  Top 10 Features  
13                     0.84       0.29  Top 10 Features  
14                     1.22      -0.48  Top 10 Features  
15                     0.86       0.26  Top 10 Features  
16                     0.93       0.13  Top 10 Features  
17                     0.88       0.24  Top 10 Features  
18                     0.79       0.37     All Features  
19                     0.82       0.33     All Features  
20                     1.17      -0.37     All Features  
21                     0.82       0.33     All Features  
22                     0.90       0.19     All Features  
23                     0.84       0.29     All Features

### Interpretation and Results

In [ ]:
XGBRegressor, _ = run_model(XGBRegressor(random_state=42), 'XGBoost Regression', X_train, y_train, X_test, y_test, verbose=False)
y_pred = XGBRegressor.predict(X_test)

# plot the predictions vs the actual values.
fig = px.scatter(x=y_test, y=y_pred, labels={'x': 'Actual', 'y': 'Predicted'})
fig.add_shape(type='line', x0=y_test.min(), y0=y_test.min(), x1=y_test.max(), y1=y_test.max())
fig.update_layout(title={'x': 0.5, 'xanchor': 'center', 'text': 'Predictions vs Actual Values'})
fig.show()


In [ ]:
# plot the residuals
fig = px.scatter(x=y_test, y=y_test - y_pred, labels={'x': 'Actual', 'y': 'Residual'})
fig.add_shape(type='line', x0=y_test.min(), y0=0, x1=y_test.max(), y1=0)
fig.update_layout(title={'x': 0.5, 'xanchor': 'center', 'text': 'Residuals'})
fig.show()


### Summary

In [ ]:
# plot a bar chart for all tbe metrics for all the models
fig = px.bar(data_frame=results, x='Model', y='Mean Absolute Error', color='Dataset', barmode='group')
fig.update_layout(title={'x': 0.5, 'xanchor': 'center', 'text': 'Mean Absolute Error'})
fig.show()

fig = px.bar(data_frame=results, x='Model', y='Root Mean Squared Error', color='Dataset', barmode='group')
fig.update_layout(title={'x': 0.5, 'xanchor': 'center', 'text': 'Root Mean Squared Error'})
fig.show()

fig = px.bar(data_frame=results, x='Model', y='R^2 Score', color='Dataset', barmode='group')
fig.update_layout(title={'x': 0.5, 'xanchor': 'center', 'text': 'R^2 Score'})
fig.show()

## Business: Insights and Recommendations

1. **Analyze busiest days and times:** According to the dataset, Saturdays have the highest frequency of orders, and most orders are placed during the night. Focus on maintaining a sufficient number of available dashers and optimizing operations during these peak times to ensure timely deliveries and high customer satisfaction.

2. **Monitor order sizes and values:** The average number of total items per order is around 3, with an average subtotal of approximately $25.63. Use this information to strategize promotions or bundle deals that cater to the typical order size and value, encouraging customers to order more frequently or spend more per order.

3. **Ensure prompt deliveries:** The average non-preparation duration (delivery time excluding the restaurant's preparation time) is around 850 seconds (14.17 minutes). Aim to keep this duration low by optimizing delivery routes and minimizing dasher idle times. Continuously monitor and evaluate delivery duration to identify areas for improvement.

4. **Maintain a balanced busy-rider ratio:** The average busy-rider ratio is 0.93, indicating that most on-shift dashers are busy with orders. Monitor this ratio to ensure a balance between the number of available
dashers and demand. A higher ratio may lead to longer wait times for customers, while a lower ratio may result in idle dashers.

5. **Focus on top-performing store categories:** The top-performing store category is "American" cuisine. Collaborate with popular restaurants in this category to develop exclusive offers and promotions, driving
customer loyalty and attracting new users to the platform.

6. **Assess delivery distance and order density:** The average delivery distance is around 2.57 units, while the average order density (orders within 10 miles of the order being processed) is approximately 0.62.
Optimize delivery routes based on these values to reduce delivery times and ensure that dashers can complete multiple orders efficiently.

7. **Offer a diverse range of cuisine options:** With 74 unique store primary categories in the dataset, it is crucial to maintain a diverse range of cuisine options to cater to various customer preferences. Collaborate with popular and emerging restaurants from different categories to expand the platform's offerings and attract a wider customer base.

8. **Optimize order protocols:** The dataset includes seven unique order protocols, with protocol 3 being the most frequent. Investigate the efficiency of different order protocols to determine if certain
protocols are more effective in reducing delivery times or streamlining operations. Implement the most efficient protocols where applicable to improve overall performance.

9. **Explore pricing strategies:** The average minimum and maximum item prices in the dataset are $6.88 and $11.49, respectively. Investigate if implementing dynamic pricing or offering promotions based on item prices can incentivize customers to order more frequently or try new items.

10. **Enhance estimated order placement and travel time accuracy:** The dataset includes estimated order placement and store-to-consumer driving durations. Continuously analyze and refine these estimates to improve
their accuracy, enabling better planning for delivery routes and dasher allocation.

11. **Analyze day-of-week and time-of-day trends:** Leverage the dataset's day-of-week and time-of-day variables to identify specific trends or patterns that can inform targeted promotions, staffing decisions, or
operational adjustments. For example, offering time-sensitive deals during off-peak hours may help increase demand and smooth out order volume throughout the day.

## Conclusion

Based on the analysis and predictions of various machine learning models for food delivery companies, the XGBoost Regressor consistently outperforms other models with a Mean Absolute Error (MAE) of 553 to 585, Root Mean Squared Error (RMSE) of 706, and R^2 Score of 0.35 to 0.37 across all feature sets. The Multi-Layer Perceptron Regressor model also performs well, but the XGBoost Regressor shows the best results in terms of error metrics and R^2 score.

## References

1.	Khiari, J., & Olaverri-Monreal, C. (n.d.). Boosting Algorithms for Delivery Time Prediction in Transportation Logistics. Papers with Code. Retrieved from https://paperswithcode.com/paper/boosting-algorithms-for-delivery-time
2.	End-to-End Prediction of Parcel Delivery Time with Deep Learning. (n.d.). arXiv. Retrieved from https://arxiv.org/pdf/2009.12197
3.	Jafari, R. (n.d.). Lecture Material. Retrieved from https://github.com/rjafari979
4.	Huang, J. (n.d.). Case Studies. Retrieved from https://github.com/yuxiaohuang/teaching
5.	Strata Scratch. (n.d.). Delivery Duration Prediction Dataset and Reference. Retrieved from https://platform.stratascratch.com/data-projects/delivery-duration-prediction
6.	Dash Gallery. (n.d.). Dash materials. Retrieved from https://dash.gallery/Portal/
